tools/fp20compiler/main.cpp

```diff
@@ -37,25 +37,69 @@ char * myin = 0;
 void translate(const char* s) {
     char* ns = strdup(s);
 
-    // printf("-- %s\b", s);
-
-    char* ts_start = strstr(ns, "!!TS1.0");
-    if (ts_start) {
-        char* ts_end = strstr(ts_start, "// End of program");
-        assert(ts_end);
-        *ts_end = 0;
-
-        ts10_init(ts_start);
-        ts10_parse();
-
-        *ts_end = '/';
-    }
-
-    char* rc_start = strstr(ns, "!!RC1.0");
-    if (rc_start) {
-        rc10_init(rc_start);
-        rc10_parse();
-    }
+    // Move to start of file
+    char* cursor = ns;
+
+    // Remember where the current shader starts
+    char* shader_start = NULL;
+
+    // Loop over each shader (between two shader start tokens)
+    do {
+
+        // Try to find a shader start token candidate to mark the end
+        char* shader_end = strstr(cursor, "!!");
+
+        // Make sure we found a token candidate
+        if (shader_end != NULL) {
+
+            // Only accept token at beginning of file, or beginning of line
+            if ((shader_end != ns) && (shader_end[-1] != '\n')) {
+                cursor = shader_end + 1;
+                continue;
+            }
+
+        }
+
+        // We need a valid start token
+        if (shader_start == NULL) {
+
+            // If we found no tokens at all, we have to give up
+            if (shader_end == NULL) {
+                fprintf(stderr, "no shaders found\n");
+                break;
+            }
+
+            // If we only found one token, use it as start + find another one
+            shader_start = shader_end;
+            cursor = shader_start + 1;
+            continue;
+        }
+
+        // Mark end of string at next token (if it exists)
+        if (shader_end) { *shader_end = '\0'; }
+
+        // Process shader
```

Comment:
This should add a comment so the user knows what generated the respective pbkit commands.

Reply:
I've added a new commit which does this.